### DISCUSSION TITLE
Customizable menus in UI

### DISCUSSION THREAD

**Discussion by innerpeacez** [apache/skywalking#10731] [Ideas]
As the skywalking features become more and more abundant, the menus displayed in the UI will become more and more numerous. However, not everyone will use all the features. At this point, these unused menus become redundant. Providing customizable configuration options for menus may make them more flexible.

My initial idea is to configure the menus through a configuration file, Users can decide which menus to keep based on their own needs, where all the menus are enabled by default. Then, an Endpoint can be provided to the UI to let it know which menus are needed.

**Comment by wu-sheng** [comment]
> My initial idea is to configure the menus through a configuration file

Actually, we talked about this, but it feels so complex for end users. 

___

The design(not one is doing this) is, the menu should be in two groups. One is the active menu, the other is available features.
About the `active` definition, it is determined by
1. It was used in the recent week(past 7 days)
2. Each menu item should bind with a metric as an indicator, which represents this menu item should be shown

___
@innerpeacez If you want to implement this, welcome to submit a design proposal

**Reply by innerpeacez** [reply]
Regarding the issue of active menus, determining whether a menu is active through metric collection seems to make the matter very complicated. Active menus will always be shown in the UI, but how can menus that have already been marked as inactive and are no longer shown be made visible again?

**Reply by innerpeacez** [reply]
Did I misunderstand your meaning about active menu?

**Reply by wu-sheng** [reply]
> determining whether a menu is active through metric collection seems to make the matter very complicated

Why it is complex? If nothing happens(no metric), there is no point to show.

> Active menus will always be shown in the UI, but how can menus that have already been marked as inactive and are no longer shown be made visible again?

Because we don't have a group(active and inactive) now. Why `always`?

**Reply by innerpeacez** [reply]
> Each menu item should bind with a metric as an indicator, which represents this menu item should be shown

About this. 

Does referring to `a metric` generated by clicking on the UI, or do `a metric` correspond to actual data collected via feature(just these metrics bind to a menu item)?

**Reply by wu-sheng** [reply]
This may not have to be `on click`. It would be a part of `init` and periodical flush.

**Comment by innerpeacez** [comment]
Active menu seems to result in users frequently seeing a different UI, which is too amazing.
